netlogo-gui/src/main/window/AutoSuggest.scala

```diff
@@ -0,0 +1,218 @@
+      if (memo.contains((s1,s2)) == false)
+        memo((s1,s2)) = (s1, s2) match {
+          case (_, Nil) => s1.length
+          case (Nil, _) => s2.length
+          case (c1::t1, c2::t2)  => min( sd(t1,s2) + 1, sd(s1,t2) + 1,
+            sd(t1,t2) + (if (c1==c2) 0 else 1) )
+        }
+      memo((s1,s2))
+    }
+
+    sd( s1.toLowerCase.toList, s2.toLowerCase.toList )
+  }
+
+  def getSuggestions(word: String): Seq[String] = {
+    val eD = if(word.length >= 2) 1 else 0
+
+    var toHitList = Seq[String]()
+    for(i <- 0 to eD) {
+      toHitList ++= trie.findByLength(word, word.length + i)
+    }
+    toHitList = for(token <- toHitList if editDistance(token, word) <= eD) yield token
+    toHitList.sortBy(editDistance(_, word))
+    var suggestionList = TreeSet[String]()
+    for(token <- toHitList) {
+      suggestionList ++= trie.findByPrefix(token)
+    }
+    suggestionList.toSeq
+      .sortBy(editDistance(_, word))
+
+  }
+
+  /**
+    * Code below contains the Trie implementation for suggestion retrieval
+    */
+  object Trie {
+    def apply() : Trie = new TrieNode()
+  }
+
+  sealed trait Trie extends Traversable[String] {
+
+    def append(key : String)
+    def findByPrefix(prefix: String): scala.collection.Seq[String]
+    def contains(word: String): Boolean
+    def remove(word : String) : Boolean
+
+  }
+
+  class TrieNode(val char : Option[Char] = None, var word: Option[String] = None) extends Trie {
+
+    val children: mutable.Map[Char, TrieNode] = new java.util.TreeMap[Char, TrieNode]().asScala
+
+    override def append(key: String) = {
+
+      @tailrec def appendHelper(node: TrieNode, currentIndex: Int): Unit = {
+        if (currentIndex == key.length) {
+          node.word = Some(key)
+        } else {
+          val char = key.charAt(currentIndex).toLower
+          val result = node.children.getOrElseUpdate(char, {
+            new TrieNode(Some(char))
+          })
+
+          appendHelper(result, currentIndex + 1)
+        }
+      }
+
+      appendHelper(this, 0)
+    }
+
+    override def foreach[U](f: String => U): Unit = {
+
+      @tailrec def foreachHelper(nodes: TrieNode*): Unit = {
+        if (nodes.size != 0) {
+          nodes.foreach(node => node.word.foreach(f))
+          foreachHelper(nodes.flatMap(node => node.children.values): _*)
+        }
+      }
+
+      foreachHelper(this)
+    }
+
+    override def findByPrefix(prefix: String): scala.collection.Seq[String] = {
+
+      @tailrec def helper(currentIndex: Int, node: TrieNode, items: ListBuffer[String]): ListBuffer[String] = {
+        if (currentIndex == prefix.length) {
+          items ++ node
+        } else {
+          node.children.get(prefix.charAt(currentIndex).toLower) match {
+            case Some(child) => helper(currentIndex + 1, child, items)
+            case None => items
+          }
+        }
+      }
+
+      helper(0, this, new ListBuffer[String]())
+    }
+
+    // Optimize this for @tailrec
```

Comment:
Are you planning to optimize this for `@tailrec`? This comment will need to be removed either way.